### DISCUSSION TITLE
welcome to ski.js discussions!

### DISCUSSION THREAD

**Discussion by saejin-moon** [thelegendski/ski.js#10] [Announcements]
## hey, :D!
  i'm usin' Discussions as a place to connect with other members of our community. i hope that you:
  * ask questions you’re wondering about.
  * share ideas.
  * engage with other community members.
  * welcome others and are open-minded. remember that this is a community we
  build together.

  To get started, comment below with an introduction of yourself and tell us about yourself an' where you use ski.js, :D.

**Comment by astro-develops** [comment]
### Interesting..... 
#### Glad to see the community getting involved with ski.js

I'm curious. For your distance function, you used the exponentiation operator, which is normal:
`
mag = (a,b)=>Math.sqrt((a ** 2) + (b ** 2))
dist = (x,y,X,Y)=>mag(x - X, y - Y)
`

but I'd like to know which is more efficient:
_The Exponentiation Operator(**) or Math.pow()?_

I'm totally clueless about this so all opinions are welcome

**Reply by saejin-moon** [reply]
neither or both. depends on how you look at it. they have the same performance. literally down to the millisecond.

oh, an' anyone that says that `x * x` is faster than `x ** 2` is wrong. they perform with the same efficiency.

**Reply by astro-develops** [reply]
I see. Kind of shocked to see they have the same efficiency.
Thanks for your opinion.

**Comment by dominicrocco** [comment]
hi ski :)

is there a link like this for the addons?
also, how to I get the latest version in this kind of tag?
`https://cdn.jsdelivr.net/gh/thelegendski/ski.js@1.4.7/ski.js`

thanks.

**Reply by saejin-moon** [reply]
yes, there is, :D.

to get latest, simply drop @ and the version number.

here's a few of the current repos you can import rn. there will be two more in the near future (1.6.3) which will be `defer.js` (creates an `OffscreenCanvas` instead. mostly for WebGL texture purposes) an' `addons/filter.js` which will filter images.

DO NOT USE OLDER VERSIONS OF ski.js! i've made so many mistakes that need fixin', added necessary features, an' improved efficiency, so you'll need to keep up to date with the new releases. my best suggestion is to simply use the link below. i'm currently on 1.6.1, but as aforementioned 1.6.3 is on its way. i wouldn't expect any major releases after that unless a major bug is found or a new feature is requested.

oh, the new file name is main.js instead of ski.js as the repo is already named ski.js, =P.

**main script**
`https://cdn.jsdelivr.net/gh/thelegendski/ski.js/main.js`
**min script [ski.js without the graphix]**
`https://cdn.jsdelivr.net/gh/thelegendski/ski.js/min.js`
**vector library** _meant to be used with `import`_
`https://cdn.jsdelivr.net/gh/thelegendski/ski.js/addons/vec.js`
**println script [now has no bugs lol]**
`https://cdn.jsdelivr.net/gh/thelegendski/ski.js/addons/print.js`
**defer will be called like this**
`https://cdn.jsdelivr.net/gh/thelegendski/ski.js/defer.js`
**image filters will be called like this**
`https://cdn.jsdelivr.net/gh/thelegendski/ski.js/addons/filters.js`

you'll find that most sources will tell you not to use the latest version of a script as it might break your program, so i'd stick with 1.6.3 once it's released within a couple weeks, :). it would be called like this.
**main script**
`https://cdn.jsdelivr.net/gh/thelegendski/ski.js@1.6.3/main.js`

**Reply by dominicrocco** [reply]
sweet, thanks. I was trying to use the version I gave you but `println` wouldn't work, :(.

i'll try this later.

**Reply by PromiseCoder** [reply]
@thelegendski The first 2 links are the same.

**Reply by saejin-moon** [reply]
that should be fixed, :p.

**Comment by dominicrocco** [comment]
question. I noticed in the vector addon you use this notation (I forgot the name, 💀).
`[this.x, this.y, this.z] = [x, y, z];`

i'm familiar with this notation, but I thought that just assigning them in three separate lines would be more efficient - i might be wrong. what's your take?

**Reply by saejin-moon** [reply]
the notation is referred to as destructuring.

when the code was written, it was to save lines of code.
after further research, it is faster by a slim margin, but still more efficient.
if they perform with the same efficiency an' one is less typin', i'm all for it, :D.

**Comment by dominicrocco** [comment]
hello,

after looking at a comment by theDark on KA i've decided to contact you about making documentation for this a reality. is there any way i can help with it?

**Reply by saejin-moon** [reply]
the reason why no documentation has been made is simply my stupid ambition. i find it necessary for the documentation to not only detail what the function does an' its parameters an' other details, but also to provide an example of what the function does. it's a dauntin' prospect, but after the landin' page an' homepage are created it shouldn't be anythin' more than tedious.

additionally, it doesn't help that i'm still not finished. once version 1.9.0 is released from the tips from TheDark an' SMV, a documentation can be created, but a template could be made anyways in the meantime.

**Reply by astro-develops** [reply]
If you'd like I can help build the documentation _=D_

**Reply by saejin-moon** [reply]
i'd be glad to get all the help i can get, :D!

y'all thinkin' now or after CoC?

**Reply by astro-develops** [reply]
I suggest creating a proper layout of the webpage rn. The boring stuff after CoC. 

Maybe you should create a ski.js team/organization?

**Reply by saejin-moon** [reply]
alr then.

done, :). invited both of you, :D.

**Comment by PromiseCoder** [comment]
### Web page input fixes
If you want to use this code, put them before your ski.js code.
The first 2 code snippets are copied directly from ski.js with the removal of `e.preventDefault()` and the addition of semicolons (necessary if you want readable code and JavaScript that runs).

Some event attributes in ski.js contain `e.preventDefault()`. This means the arrow keys won't make the page scroll. However, it also disables `<input>` and `<textarea>` editing and keyboard shortcuts for the rest of the web page.
Removing `e.preventDefault()` will fix this (but I could be unaware of some other consequence).
```
document.onkeydown = e => {
    key = e.key;
    keyCode = e.keyCode;
    keyIsPressed = true;
    keyPressed(e);
};
```

Similarly, removing `e.preventDefault()` from `canvas.onwheel` will turn off the scroll lock when your cursor is over the ski.js canvas.
```
canvas.onwheel = e => {
    mouseScrolled(e);
};
```
or
```
canvas.onwheel = e => mouseScrolled(e);
```

If you don't need keys or scroll input from ski.js (say, if you only use it as part of thumbnail.js for a web page), these work as well.
```
document.onkeydown = null;
canvas.onwheel = null;
```

**Reply by saejin-moon** [reply]
i think i fixed it, :P. check v1.1.0.

**Reply by PromiseCoder** [reply]
cool. :D
should I delete this, do ya think?

**Reply by saejin-moon** [reply]
idrc, :P.
